### PR TITLE
sink(ticdc): store checkpointTs by flushedResolvedTsMap in mysql sink

### LINKED ISSUE #5107: consistent_replicate_gbk losing data after applying redolog
### Which jobs are flaking?

[cdc_ghpr_integration_test](https://ci.pingcap.net/blue/organizations/jenkins/cdc_ghpr_integration_test/activity)[](https://ci.pingcap.net/blue/organizations/jenkins/cdc_ghpr_integration_test/detail/cdc_ghpr_integration_test/2595/pipeline)

### Which test(s) are flaking?

consistent_replicate_gbk

### Jenkins logs or GitHub Actions link

```

[2022-04-01T16:36:43.146Z] [2022/04/02 00:36:43.133 +08:00] [INFO] [redo.go:103] ["apply redo log starts"] [checkpointTs=432231149290913799] [resolvedTs=432231153485217794]

[2022-04-01T16:36:43.146Z] [2022/04/02 00:36:43.141 +08:00] [INFO] [mysql_params.go:314] ["sink uri is configured"] [dsn="normal:******@tcp(127.0.0.1:3306)/?interpolateParams=true&multiStatements=true&allow_auto_random_explicit_insert=1&charset=utf8mb4&readTimeout=2m&sql_mode=%22IGNORE_SPACE%2CALLOW_INVALID_DATES%2CNO_ENGINE_SUBSTITUTION%2CONLY_FULL_GROUP_BY%2CNO_AUTO_VALUE_ON_ZERO%22&tidb_placement_mode=%22ignore%22&tidb_txn_mode=optimistic&time_zone=%22UTC%22&timeout=2m&transaction_isolation=%22READ-COMMITTED%22&writeTimeout=2m"]

[2022-04-01T16:36:43.428Z] [2022/04/02 00:36:43.145 +08:00] [INFO] [mysql.go:180] ["Start mysql sink"]

[2022-04-01T16:36:43.428Z] Apply redo log successfully

[2022-04-01T16:36:43.693Z] check diff failed 1-th time, retry later

[2022-04-01T16:36:46.316Z] check diff failed 2-th time, retry later

[2022-04-01T16:36:48.280Z] check diff failed 3-th time, retry later

[2022-04-01T16:36:50.226Z] check diff failed 4-th time, retry later

[2022-04-01T16:36:52.848Z] check diff failed 5-th time, retry later

[2022-04-01T16:36:54.775Z] check diff failed 6-th time, retry later

[2022-04-01T16:36:57.353Z] check diff failed 7-th time, retry later

[2022-04-01T16:36:59.333Z] check diff failed 8-th time, retry later

[2022-04-01T16:37:01.921Z] check diff failed 9-th time, retry later
```

https://ci.pingcap.net/blue/organizations/jenkins/cdc_ghpr_integration_test/detail/cdc_ghpr_integration_test/2595/pipeline/

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #5107

### What is changed and how it works?
- When mysql Sink traverses the`tableMaxResolvedTs ` at this [line](https://github.com/pingcap/tiflow/blob/f3bf091a656301a59f2c98fff0ed5e49355d49e8/cdc/sink/mysql/mysql.go#L272), the map may have been concurrently modified, [which will result in an incorrect value being saved to`tableCheckpointTs `](https://github.com/pingcap/tiflow/blob/f3bf091a656301a59f2c98fff0ed5e49355d49e8/cdc/sink/mysql/mysql.go#L273).
To solve this problem, we need to: 
    1. Save the resolvedTs, which are used to [search data](https://github.com/pingcap/tiflow/pull/5418/files#diff-ab73e67a0b47a3a9537743cf306c53bbf2247b8b53e8a1998ac3529e4c2d41b9R145) from txnCache, into the [flushedResolvedTsMap](https://github.com/pingcap/tiflow/pull/5418/files#diff-ab73e67a0b47a3a9537743cf306c53bbf2247b8b53e8a1998ac3529e4c2d41b9R130)
    2. In mysql sink, always store checkpointTs based on [flushedResolvedTsMap](https://github.com/pingcap/tiflow/pull/5418/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R282)

- Replace `resolvedNotifier` with `resolvedCh` to simplify the code

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
`Fix a bug that mysql sink may save a wrong checkpointTs`.
```

### CODE REVIEW

File: cdc/sink/mysql/txn_cache.go [hunk 1]
```diff
@@ -115,48 +115,53 @@ func (c *unresolvedTxnCache) Resolved(
...
-		if !ok {
+	resolvedTsMap.Range(func(k, v any) bool {
+		tableID := k.(model.TableID)
+		resolvedTs := v.(model.Ts)
+		flushedResolvedTsMap[tableID] = resolvedTs
+		return true
+	})
+
+	resolvedRowsMap = make(map[model.TableID][]*model.SingleTableTxn, len(unresolvedTxns))
+	var (
```
Comment (zhaoxinyu) [inline cdc/sink/mysql/txn_cache.go:134]: It's better to put these variable definitions at the head of function body.

**Review by zhaoxinyu — APPROVED** [review]

File: cdc/sink/mysql/mysql.go [hunk 2]
```diff
@@ -533,7 +524,7 @@ func (s *mysqlSink) cleanTableResource(tableID model.TableID) {
 
 func (s *mysqlSink) Close(ctx context.Context) error {
 	s.execWaitNotifier.Close()
-	s.resolvedNotifier.Close()
+	close(s.resolvedCh)
```
Comment (liuzix) [inline cdc/sink/mysql/mysql.go:527]: Why is it necessary to close `s.resolvedCh`?
Can `Close(...)` race with `FlushRowChangedEvents(...)`? If so, `FlushRowChangedEvents` will panic if it tries to send to a closed channel.

Comment (CharlesCheung96) [reply cdc/sink/mysql/mysql.go]: Yes, there are data race problem, thank you for your review

**Review by liuzix — APPROVED** [review]
LGTM

**Pull request merged**